Add Symbolic Shape Registry and define in C++

eellison opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #58932 Add more functions
* #58890 Switch symbolic shape registry to operator map'
* #56966 [JIT] Disable Complete Shape Inlining For Testing Purposes
* **#56815 Add Symbolic Shape Registry and define in C++**
* #56828 Add unary/binary ops necessary for mobilenet

Differential Revision: [D28444662](https://our.internmc.facebook.com/intern/diff/D28444662)

penguinwu commented on torch/csrc/jit/passes/symbolic_shape_analysis.cpp:
```diff
@@ -270,24 +268,12 @@ void PropagateShapesWithShapeFunction(
-  operator_functions[toString(n->schema())] = graph;
-}
-
 void PropagateShapesOnGraph(std::shared_ptr<Graph>& graph) {
```
Could you add a comment on what `PropagateShapesOnGraph` does? It is not clear to me what we are propagating. I would expect some input shapes to be passed into `PropagateShapesOnGraph` for propagation, but the code seems to just propagate the symbolic shapes (represented as `shapeComputeGraphForSchema`)  associated w/ the node to compute the shape of the output nodes? If we do not use any outside information, then it seems that we could statically deduce the output tensor shapes. I could have misunderstood the codes, but it is definitely not intuitive and warrants some comments here.

penguinwu commented on torch/csrc/jit/runtime/symbolic_shape_registry.cpp:
```diff
@@ -0,0 +1,130 @@
+// TODO: if a unary function has an equivalent inplace operation, then the shape function must 
+// be an identity. We could add those automatically. 
+
+std::unordered_map<std::string,  std::string> schema_string_to_function_name_mappings = {
```
Reaability nip: `schema_string_to_function_name_mappings` -> `schema_string_to_shape_function_name_mappings`

penguinwu commented on torch/csrc/jit/runtime/symbolic_shape_registry.cpp:
```diff
@@ -0,0 +1,130 @@
+    {"aten::adaptive_avg_pool2d(Tensor self, int[2] output_size) -> (Tensor)", "adaptive_avg_pool2d"},
+};
+
+std::unordered_map<std::string, std::shared_ptr<Graph>> schema_string_to_function_graph;
```
Readability: `schema_string_to_function_graph` -> `schema_string_to_shape_function_graph`

penguinwu commented on torch/csrc/jit/runtime/symbolic_shape_registry.cpp:
```diff
@@ -0,0 +1,130 @@
+
+std::unordered_map<std::string, std::shared_ptr<Graph>> schema_string_to_function_graph;
+
+std::unordered_map<const FunctionSchema*, std::shared_ptr<Graph>> cached_schema_to_graph;
```
Readability: `cached_schema_to_graph` -> `cached_schema_to_shape_function_graph`

penguinwu commented on torch/csrc/jit/runtime/symbolic_shape_registry.cpp:
```diff
@@ -0,0 +1,130 @@
+
+
+void loadModule(const CompilationUnit& module) {
+  std::unordered_map<std::string, std::shared_ptr<Graph>> reused_functions;
```
[Readability nip] It is easier to understand to use `shape_function` instead of `function` in the following codes.

penguinwu reviewed (changes requested):
LGTM, just some readability nips and one question regarding `PropagateShapesOnGraph` (that may require additional comment)

eellison commented:
@eellison has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D28444662).

eellison commented:
@eellison has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D28444662).

eellison commented:
@eellison has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D28444662).

eellison commented:
@eellison has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D28444662).

eellison commented:
@eellison has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D28444662).

eellison committed:
Update on "Add Symbolic Shape Registry and define in C++"


Differential Revision: [D28444662](https://our.internmc.facebook.com/intern/diff/D28444662)

[ghstack-poisoned]

eellison commented:
@eellison has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D28444662).

eellison committed:
Update on "Add Symbolic Shape Registry and define in C++"


Differential Revision: [D28444662](https://our.internmc.facebook.com/intern/diff/D28444662)

[ghstack-poisoned]
